File: pokemon_go_calc/models/iv.py
"""個体値モデル."""

from pydantic import BaseModel, Field


class IV(BaseModel):
    """個体値（Individual Values）.

    Attributes:
        attack: 攻撃個体値（0-15）
        defense: 防御個体値（0-15）
        stamina: HP個体値（0-15）
    """

    attack: int = Field(ge=0, le=15, description="攻撃個体値")
    defense: int = Field(ge=0, le=15, description="防御個体値")
    stamina: int = Field(ge=0, le=15, description="HP個体値")

    model_config = {"frozen": True}

    def __str__(self) -> str:
        """文字列表現を返す."""
        return f"{self.attack}/{self.defense}/{self.stamina}"

    @property
    def total(self) -> int:
        """合計値を返す."""
        return self.attack + self.defense + self.stamina

    @property
    def percentage(self) -> float:
        """パーセンテージを返す（0-100%）."""
        return (self.total / 45) * 100

File: linebot/src/Domain/ValueObjects/IV.php
<?php

declare(strict_types=1);

namespace PokemonGoCalc\LineBot\Domain\ValueObjects;

/**
 * 個体値 (Individual Values) Value Object.
 */
final class IV
{
    private const MIN = 0;
    private const MAX = 15;

    public function __construct(
        public readonly int $attack,
        public readonly int $defense,
        public readonly int $stamina,
    ) {
        $this->validate($attack, 'attack');
        $this->validate($defense, 'defense');
        $this->validate($stamina, 'stamina');
    }

    private function validate(int $value, string $name): void
    {
        if ($value < self::MIN || $value > self::MAX) {
            throw new \InvalidArgumentException(
                sprintf('%s must be between %d and %d', $name, self::MIN, self::MAX)
            );
        }
    }

    public function __toString(): string
    {
        return sprintf('%d/%d/%d', $this->attack, $this->defense, $this->stamina);
    }

    public function total(): int
    {
        return $this->attack + $this->defense + $this->stamina;
    }

    public function percentage(): float
    {
        return ($this->total() / (self::MAX * 3)) * 100;
    }

    public function isPerfect(): bool
    {
        return $this->attack === self::MAX
            && $this->defense === self::MAX
            && $this->stamina === self::MAX;
    }
}
